evm/v0.5/contracts/PreCoordinator.sol

```diff
@@ -0,0 +1,295 @@
+    saId = keccak256(abi.encodePacked(globalNonce, now));
+    globalNonce++; // yes, let it overflow
+    // Manually calculate total payment off-chain
+    serviceAgreements[saId] = ServiceAgreement(_totalPayment, _minResponses, _oracles, _jobIds, _payments);
+
+    emit NewServiceAgreement(saId, _totalPayment, _minResponses);
+  }
+
+  /**
+   * @notice This is a helper function to retrieve the details of a service agreement
+   * by its given service agreement ID.
+   * @dev This function is used instead of the public mapping to return the values
+   * of the arrays: oracles, jobIds, and payments.
+   */
+  function getServiceAgreement(bytes32 _saId)
+    external view returns
+  (
+    uint256 totalPayment,
+    uint256 minResponses,
+    address[] memory oracles,
+    bytes32[] memory jobIds,
+    uint256[] memory payments
+  )
+  {
+    return
+    (
+      serviceAgreements[_saId].totalPayment,
+      serviceAgreements[_saId].minResponses,
+      serviceAgreements[_saId].oracles,
+      serviceAgreements[_saId].jobIds,
+      serviceAgreements[_saId].payments
+    );
+  }
+
+  /**
+   * @notice Deletes a service agreement from storage
+   * @dev Use this with caution since there may be responses waiting to come
+   * back for a service agreement. This can be monitored off-chain by looking
+   * for the ServiceAgreementRequest event.
+   * @param _saId The service agreement ID
+   */
+  function deleteServiceAgreement(bytes32 _saId) external onlyOwner {
+    delete serviceAgreements[_saId];
+  }
+
+  /**
+   * @notice Returns the address of the LINK token
+   * @dev This is the public implementation for chainlinkTokenAddress, which is
+   * an internal method of the ChainlinkClient contract
+   */
+  function getChainlinkToken() public view returns (address) {
+    return chainlinkTokenAddress();
+  }
+
+  /**
+   * @notice Creates the Chainlink request
+   * @dev Stores the hash of the params as the on-chain commitment for the request.
+   * Emits OracleRequest event for the Chainlink node to detect.
+   * @param _sender The sender of the request
+   * @param _payment The amount of payment given (specified in wei)
+   * @param _saId The Job Specification ID
+   * @param _callbackAddress The callback address for the response
+   * @param _callbackFunctionId The callback function ID for the response
+   * @param _nonce The nonce sent by the requester
+   * @param _data The CBOR payload of the request
+   */
+  function oracleRequest(
+    address _sender,
+    uint256 _payment,
+    bytes32 _saId,
+    address _callbackAddress,
+    bytes4 _callbackFunctionId,
+    uint256 _nonce,
+    uint256,
+    bytes calldata _data
+  )
+    external
+    onlyLINK
+    checkCallbackAddress(_callbackAddress)
+  {
+    ServiceAgreement memory sa = serviceAgreements[_saId];
+    // this revert message does not bubble up
+    require(_payment >= sa.totalPayment, "Insufficient payment");
```

Comment:
Gas savings is applied now. However, we can't throw this into a modifier (at least not efficiently) because we need to pull the totalPayment from `serviceAgreements[_saId].totalPayment`, then check that against the `_payment` param. If we were to do this in a modifier, we'd need to read state twice.

As far as I can tell, `delegatecall` is preventing the revert message from bubbling up. This is probably something that we can remove with `abi.decode` in a later PR.